crates/ty_python_semantic/src/types/builder.rs

```diff
@@ -793,7 +793,9 @@ impl<'db> InnerIntersectionBuilder<'db> {
 
                 for existing_positive in &self.positive {
                     // S & ~T = Never    if S <: T
-                    if existing_positive.is_subtype_of(db, new_negative) {
+                    if existing_positive.is_subtype_of(db, new_negative)
+                        || existing_positive.is_gradual_equivalent_to(db, new_negative)
```

Comment:
I think the correct fix might be to use `is_assignable_to` here instead of `is_subtype_of`?

Reply:
No, we shouldn't simplify intersections according to assignability -- that would imply that e.g. `int & Any` collapses, which it should not. `Any` could materialize to either a wider or narrower type than `int`, so it doesn't reduce (it represents a dynamic type with upper bound of `int`.)